cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,342 @@
+However, the specification does not otherwise specify or limit the space of expressible constraints that the syntax and semantics allow.
+
+=== Syntax
+
+The constraint syntax is defined as follows:
+
+.Grammar definition for constraint syntax.
+[source, ebnf]
+----
+constraint command    = create-constraint | drop-constraint ;
+create-constraint     = "CREATE", "CONSTRAINT", [ constraint-name ], "FOR", pattern, "REQUIRE", constraint-predicate, { "REQUIRE", constraint-predicate } ;
+constraint-name       = symbolic-name
+constraint-predicate  = expression | unique | primary-key ;
+unique                = "UNIQUE", property-expression
+primary-key           = "PRIMARY KEY", property-expression, { ",", property-expression }
+drop-constraint       = "DROP", "CONSTRAINT", constraint-name ;
+----
+
+The `REQUIRE` clause works exactly like the `WHERE` clause in a standard Cypher query, with the addition of also supporting the special constraint operators `UNIQUE` and `PRIMARY KEY`.
+This allows for very complex concrete constraint definitions (using custom predicates) within the specified syntax.
+
+For details on `UNIQUE` and `PRIMARY KEY`, see the dedicated sections below: <<uniqueness>>, <<primary-key>>.
+
+==== Constraint names
+
+All constraints provide the user the option to specify a nonempty _name_ at constraint creation time.
+This name is subsequently the handle with which a user may refer to the constraint, for example when dropping it.
+In the case where a name is not provided, the system will generate a unique name.
+
+==== Removing constraints
+
+A constraint is removed by referring to its name.
+
+.Example of dropping a constraint with name `foo`:
+[source, cypher]
+----
+DROP CONSTRAINT foo
+----
+
+=== Semantics
+
+The semantics for constraints follow these general rules:
+
+1. The constraint pattern define the constraint _domain_, where all entities that would be returned by a `MATCH` clause with the same pattern constitute the domain, with one notable exception (see <<domain-exception, 3.>>).
+
+2. The constraint expressions defined in the `REQUIRE` clauses of the constraint definition must all evaluate to `true`.
+
+3. [[domain-exception]]Entities for which a constraint expression evaluate to `null` under Cypher's ternary logic are _excluded_ from the constraint domain, even if they fit within the constraint pattern.
+
+==== Errors
+
+The following list describes the situations in which an error will be raised:
+
+* Attempting to create a constraint on a graph where the data does not comply with the constraint criterion.
+* Attempting to create a constraint with a name that already exists.
+* Attempting to drop a constraint referencing a non-existent name.
+* Attempting to modify the graph in such a way that it would violate a constraint.
+
+==== Mutability
+
+Once a constraint has been created, it may not be amended.
+Should a user wish to change its definition, it has to be dropped and recreated with an updated structure.
```

Comment:
Should we have a note here that transactional implementations could do both the dropping and recreation in the same transaction so that the constraint is atomically mutated? This would of course allow leaving the old constraint in place should the creation of the new constraint fail.